### PR TITLE
Feature/binary UUID

### PR DESCRIPTION
**Pull request by bradennapier**
### Pull Request check-list

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

#10265 #9271 #4997 

This **starts** to implement a binary uuid type.  It is meant to open the general discussion around what may be missing.  I have implemented a working package as a module here to give a better idea: https://github.com/odo-network/sequelize-binary-uuid

The benefit is fairly massive (@see: https://www.percona.com/blog/2014/12/19/store-uuid-optimized-way/)

This PR implements the `UUIDV1` shuffling as recommended by above as well as mysql team directly to greatly improve performance while indexing UUIDs.

### Example

```javascript
const User = sequelize.define(
  'TableName',
  {
    id: {
      primaryKey: true,
      type: Sequelize.DataTypes.BINARYUUID,
      defaultValue: Sequelize.DataTypes.UUIDV1.BINARY
    }
  }
);
```

image

---

### `DataTypes.UUID.BINARY`

I could not make it work when trying to use `UUID.BINARY` as it was breaking when going to dialects and not passing options on, among other things... probably makes the most sense there but way too many side effects so I changed to using `BINARYUUID` type.

### Virtual / Transformed Field

In https://github.com/odo-network/sequelize-binary-uuid it also provides a virtual representation of the field in a fairly straight-forward way.  I wasn't sure the best way to do something similar here.

https://github.com/odo-network/sequelize-binary-uuid#using-withbinaryuuid-preset-recommended
https://github.com/odo-network/sequelize-binary-uuid#creating-a-uuid-virtual-field

Specifically since the `uuidv1` is transformed to binary - there should be a simple and easy way to view the field as its non-binary form.  

### v5 breaks `X'${hex}'` handling

It would appear that v5 breaks the handling of anything using `X'...` - I haven't tested but as far as I can tell this would break a few areas such as `BLOB` but I may be wrong.  In the end it seems that stringify just doesn't make sense to use.

I was hoping it would work though as it makes the LOGGING much more helpful than just putting up question marks like it does now.

### CODE REVIEW

**Comment by bradennapier** [comment]
I am not really sure why MACADDR is being parsed as binary(16) - they have nothing to do with each other at all... so will need some clarification as-to why it thinks that.

File: lib/data-types.js [hunk 1]
```diff
@@ -789,14 +790,38 @@ UUID.prototype.validate = function validate(value, options) {
   return true;
 };
 
+/**
+ * A column storing a modified uuidv1 as binary(16).
+ * @see https://www.percona.com/blog/2014/12/19/store-uuid-optimized-way/
+ *
+ * @namespace DataTypes.BINARYUUID
+ */
+function BINARYUUID() {
```
Comment (SimonSchick) [inline lib/data-types.js]: @sushantdhiman I was wondering if you want to enforce these to be classes (and be called with `new`) in v5?

File: lib/data-types.js [hunk 2]
```diff
@@ -789,14 +790,38 @@ UUID.prototype.validate = function validate(value, options) {
...
+
 /**
  * A default unique universal identifier generated following the UUID v1 standard
  *
+ * @param {boolean} [binary=false] Whether UUID should be transformed to binary(16)
  * @namespace DataTypes.UUIDV1
  */
-function UUIDV1() {
-  if (!(this instanceof UUIDV1)) return new UUIDV1();
+function UUIDV1(binary) {
```
Comment (SimonSchick) [inline lib/data-types.js]: I suggest you make this parameter an object instead, avoid future breaking changes if we wish to add more optional parameters.

File: lib/utils.js [hunk 3]
```diff
@@ -8,7 +8,7 @@ const uuidv4 = require('uuid/v4');
 const Promise = require('./promise');
 const operators  = require('./operators');
 const operatorsArray = _.values(operators);
-
+const toBinaryUUID = require('binary-uuid').toBinaryUUID;
```
Comment (SimonSchick) [inline lib/utils.js]: Is it really worth it adding yet another dependency for a very niche data type like this?
Wouldn't it make more sense to split special datatypes and dialects into smaller packages?

File: lib/data-types.js (hunk 1 above)
Comment (bradennapier) [reply lib/data-types.js]: I'd prefer it!  Was just following what I saw previous!

File: lib/utils.js (hunk 3 above)
Comment (bradennapier) [reply lib/utils.js]: The dependency was written for this PR specifically - just left it as its own package as it was small and could be used in other places easily.

**Comment by sushantdhiman** [comment]
For now let's keep this as a separate package @bradennapier , if there is more demand we will add its support to Sequelize

**Comment by omidkrad** [comment]
@sushantdhiman Yes, there is demand. Thanks!

**Comment by NickCoad-NewsPerform** [comment]
I'm surprised this was called a niche use case. This package implements UUIDs in the way they're supposed to be implemented when used for primary keys, and UUIDs as primary keys are certainly not niche or uncommon.

Another +1 here, this would be invaluable for us.

**Comment by bradennapier** [comment]
In the meantime, the package does work well :) - but agreed it should be built in.

**Comment by Seth10001** [comment]
+1 for me too, I would like to see this included!

**Pull request closed without merging**